torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -487,9 +504,12 @@ void ProcessGroupAgent::listenLoop() {
 }
 
 void ProcessGroupAgent::pollTimedOutRPCs() {
-  while (!shutdown_.load()) {
-    std::chrono::milliseconds sleepTime;
+  while (true) {
     std::unique_lock<std::mutex> lock{futureMutex_};
+    if (!rpcRunning_.load()) {
+      return;
+    }
+    std::chrono::milliseconds sleepTime;
```

Comment:
I am not convinced yet. How do you make sure the notification is always delivered with no cond var predicate added?

e.g. What if `pollTimedOutRPCs` is at here, and `shutdown` sends a notification? Would that notification be lost?

Reply:
For `shutdown` to have sent a notification, this implies that it would have had to hold the `futureMutex` lock and set `rpcRunning` to be false before. But since that now requires the lock, we can't have `shutdown` be at `std::chrono::milliseconds sleepTime;` at the same time, since this section is also locked.